docs/source/format/CDataInterface.rst

```diff
@@ -0,0 +1,777 @@
+  children have names ``ints`` and ``floats``, and format strings ``i`` and
+  ``f`` respectively.
+* A ``map<string, float64>`` array has format string ``+m``; its single child
+  has name ``entries`` and format string ``+s``; its two grandchildren have names
+  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
+* A ``sparse_union<ints: int32, floats: float32>`` with type ids ``4, 5``
+  has format string ``+us:4,5``; its two children have names ``ints`` and
+  ``floats``, and format strings ``i`` and ``f`` respectively.
+
+
+Structure definitions
+=====================
+
+The following free-standing definitions are enough to support the Arrow
+C data interface in your project.  Like the rest of the Arrow project, they
+are available under the Apache License 2.0.
+
+.. code-block:: c
+
+   #define ARROW_FLAG_DICTIONARY_ORDERED 1
+   #define ARROW_FLAG_NULLABLE 2
+   #define ARROW_FLAG_MAP_KEYS_SORTED 4
+
+   struct ArrowArray {
+     // Type description
+     const char* format;
+     const char* name;
+     const char* metadata;
+     int64_t flags;
+
+     // Data description
+     int64_t length;
+     int64_t null_count;
+     int64_t offset;
+     int64_t n_buffers;
+     int64_t n_children;
+     const void** buffers;
+     struct ArrowArray** children;
+     struct ArrowArray* dictionary;
+
+     // Release callback
+     void (*release)(struct ArrowArray*);
+     // Opaque producer-specific data
+     void* private_data;
+   };
+
+The ArrowArray structure
+------------------------
+
+.. c:member:: const char* ArrowArray.format
+
+   Mandatory.  A null-terminated, UTF8-encoded string describing
+   the data type.  If the data type is nested, child types are not
+   encoded here but in the :c:member:`ArrowArray.children` arrays.
+
+   Consumers MAY decide not to support all data types, but they
+   should document this limitation.
+
+.. c:member:: const char* ArrowArray.name
+
+   Optional.  A null-terminated, UTF8-encoded string of the field
+   or array name.  This is mainly used to reconstruct child fields
+   of nested arrays.
+
+   Producers MAY decide not to provide this information, and consumers
+   MAY decide to ignore it.  If omitted, MAY be NULL or an empty string.
+
+.. c:member:: const char* ArrowArray.metadata
+
+   Optional.  A binary string describing the type's metadata.
+   If the data type is nested, child types are not encoded here but
+   in the :c:member:`ArrowArray.children` arrays.
+
+   This string is not null-terminated but follows a specific format::
+
+      int32: number of key/value pairs (noted N below)
+      int32: byte length of key 0
+      key 0 (not null-terminated)
+      int32: byte length of value 0
+      value 0 (not null-terminated)
+      ...
+      int32: byte length of key N - 1
+      key N - 1 (not null-terminated)
+      int32: byte length of value N - 1
+      value N - 1 (not null-terminated)
+
+   Integers are stored little-endian.  For example, the metadata
```

Comment:
nit: since this is meant for local machine transmission, should this simply be machine endianness?  I thought persisting this representation was a non-goal?

Reply:
Ah, right, endianness can simply be native. Good catch :-)